datafusion/physical-expr/src/window/aggregate.rs

```diff
@@ -98,70 +125,238 @@ impl AggregateWindowExpr {
+    ) -> Result<ArrayRef> {
+        self.accumulator.update_batch(value_slice)?;
+        let value = self.accumulator.evaluate()?;
+        Ok(value.to_array_of_size(len))
+    }
+
+    /// It calculates the running window logic.
+    /// We iterate each row to calculate its corresponding window. It is a running
+    /// window calculation. First, cur_range is calculated, then it is compared with last_range.
+    /// We increment the accumulator by update and retract.
+    /// Note that not all aggregators implement retract_batch just yet.
+    fn calculate_running_window(
+        &mut self,
+        value_slice: &[ArrayRef],
+        order_bys: &[&ArrayRef],
+        value_range: &Range<usize>,
+    ) -> Result<ArrayRef> {
+        let len = value_range.end - value_range.start;
+        let slice_order_columns = order_bys
+            .iter()
+            .map(|v| v.slice(value_range.start, value_range.end - value_range.start))
+            .collect::<Vec<_>>();
+
+        let updated_zero_offset_value_range = Range {
+            start: 0,
+            end: value_range.end - value_range.start,
+        };
+        let mut row_wise_results = vec![];
+        let mut last_range: (usize, usize) = (
+            updated_zero_offset_value_range.start,
+            updated_zero_offset_value_range.start,
+        );
+
+        for i in 0..len {
+            let window_frame = self.window_frame.ok_or_else(|| {
+                DataFusionError::Internal(
+                    "Window frame cannot be empty to calculate window ranges".to_string(),
+                )
+            })?;
+            let cur_range =
+                calculate_current_window(window_frame, &slice_order_columns, len, i)?;
+
+            if cur_range.1 - cur_range.0 == 0 {
+                // We produce None if the window is empty.
+                row_wise_results.push(ScalarValue::try_from(self.field.data_type()))
+            } else {
+                let update: Vec<ArrayRef> = value_slice
+                    .iter()
+                    .map(|v| v.slice(last_range.1, cur_range.1 - last_range.1))
+                    .collect();
+                let retract: Vec<ArrayRef> = value_slice
+                    .iter()
+                    .map(|v| v.slice(last_range.0, cur_range.0 - last_range.0))
+                    .collect();
+                self.accumulator.update_batch(&update)?;
+                // Prevents error raising if retract is not implemented.
+                if cur_range.0 - last_range.0 > 0 {
+                    self.accumulator.retract_batch(&retract)?
+                }
+                match self.accumulator.evaluate() {
+                    Err(_e) => row_wise_results
+                        .push(ScalarValue::try_from(self.field.data_type())),
+                    value => row_wise_results.push(value),
+                }
+            }
+            last_range = cur_range;
+        }
+        let row_wise_results: Result<Vec<ScalarValue>> =
```

Comment:
since we are iterating through the vec anyways above, rather than incrementally gathering `Vec<Result<_>>` you could use `?` and only push to `row_wise_results` on success. It might make this code slightly simpler